src/Chainweb/Rosetta/Util.hs

```diff
@@ -0,0 +1,686 @@
+  | coinV2Upgrade v cid bheight = _remediationMatchFunction r
+  | otherwise = _nonGenesisMatchFunction f
+  where
+    bheight = _blockHeight bh
+    cid = _blockChainId bh
+    v = _blockChainwebVersion bh
+
+
+---------------------
+-- Genesis Helpers --
+---------------------
+
+-- |
+-- | Genesis transactions do not have coinbase or gas payments.
+-- |
+getGenesisLog
+    :: Map TxId [AccountLog]
+    -> CommandResult Hash
+    -> Transaction
+getGenesisLog logs cr =
+  case (_crTxId cr) of
+    Just tid -> case (M.lookup tid logs) of
+      Just l -> rosettaTransaction cr $ makeOps tid l
+      Nothing -> rosettaTransaction cr []  -- not a coin contract tx
+    Nothing -> rosettaTransaction cr [] -- all genesis tx should have a txid
+  where
+    makeOps tid l = indexedOperations $
+      map (operation Successful TransferOrCreateAcct tid) l
+
+
+-- |
+-- | Matches all genesis transactions to their coin contract logs.
+-- |
+genesisTransactions
+    :: Map TxId [AccountLog]
+    -> V.Vector (CommandResult Hash)
+    -> GenesisMatchFunction [Transaction]
+genesisTransactions logs txs =
+  GenesisMatchFunction $
+  pure $ V.toList $ V.map (getGenesisLog logs) txs
+
+
+-- |
+-- | Matches a single genesis transaction to its coin contract logs.
+-- |
+genesisTransaction
+    :: Map TxId [AccountLog]
+    -> V.Vector (CommandResult Hash)
+    -> RequestKey
+    -- ^ target tx
+    -> GenesisMatchFunction Transaction
+genesisTransaction logs rest target = GenesisMatchFunction $ do
+  cr <- note RosettaTxIdNotFound $
+        V.find (\c -> (_crReqKey c) == target) rest
+  pure $ getGenesisLog logs cr
+
+
+------------------------
+-- Coinbase Helpers --
+------------------------
+
+-- | Matches the first coin contract logs to the coinbase tx
+nonGenesisCoinbase
+    :: (TxId, [AccountLog])
+    -> (a -> Maybe TxId)
+    -> (a -> [Operation] -> b)
+    -> a
+    -> Either String b
+nonGenesisCoinbase (tid,l) getTxId f cr = do
+  expectedTid <- note "No Coinbase TxId found" (getTxId cr)
+  if (expectedTid == tid)
+    then let ops = indexedOperations $
+                   map (operation Successful CoinbaseReward tid) l
+         in pure $ (f cr ops)
+    else Left "First log's txId does not match coinbase tx's TxId"
+
+
+------------------------
+-- NonGenesis Helpers --
+------------------------
+
+-- | Algorithm:
+--   Assumes logs at current index (i.e. idx) funded the tx.
+--   Attempts to get the coin contract logs the tx
+--   might have. If the tx succeeded (i.e. the tx has a TxId),
+--   it peeks at the next logs in the list (i.e. idx + 1).
+--   If the log's txId matches the tx's txId, then these
+--   logs are associated with the tx and the logs that
+--   paid for gas are retrieved from following index (i.e. idx + 1 + 1).
+--   If the txIds don't match OR if the tx failed, the gas logs
+--   are retrieved from idx + 1 instead.
+-- |
+gasTransactionAcc
+    :: V.Vector (TxId, [AccountLog])
+    -> (a -> Maybe TxId)
```

Comment:
Again wondering about indirection here

Reply:
Also it's bewildering trying to understand what roles `a`, `b` and `c` have. Please consider specializing to concrete types for legibility, and if not, use type variables that give some idea what is going on.

Reply:
See comment: https://github.com/kadena-io/chainweb-node/pull/1050#discussion_r440313717